components/Common/CodeBox/index.tsx

```diff
@@ -0,0 +1,107 @@
+import {
+  DocumentDuplicateIcon,
+  CodeBracketIcon,
+} from '@heroicons/react/24/outline';
+import type { FC, PropsWithChildren, ReactNode } from 'react';
+import { Fragment, isValidElement, useMemo, useRef } from 'react';
+import { FormattedMessage } from 'react-intl';
+
+import Button from '@/components/Common/Button';
+import { useCopyToClipboard } from '@/hooks/useCopyToClipboard';
+import { useNotification } from '@/hooks/useNotification';
+import { getCodeLanguageDisplayName } from '@/util/getCodeLanguageDisplayName';
+
+import styles from './index.module.css';
+
+type CodeBoxProps = {
+  language: string;
+};
+
+export const CodeBox: FC<PropsWithChildren<CodeBoxProps>> = ({
+  children,
+  language,
+}) => {
+  const ref = useRef<HTMLPreElement>(null);
+
+  const notify = useNotification();
+  const [, copyToClipboard] = useCopyToClipboard();
+
+  const onCopy = () => {
+    if (ref.current !== null && ref.current.textContent !== null) {
+      copyToClipboard(ref.current.textContent);
+
+      notify({
+        duration: 3000,
+        message: (
+          <div className={styles.notification}>
+            <CodeBracketIcon className={styles.icon} />
+            <FormattedMessage id="components.common.codebox.copied" />
+          </div>
+        ),
+      });
+    }
+  };
+
+  const code = useMemo(() => transformCode(children), [children]);
+
+  return (
+    <div className={styles.root}>
+      <pre ref={ref} className={styles.content}>
+        {code}
+      </pre>
+      <div className={styles.footer}>
+        <span className={styles.language}>{language}</span>
+        <Button type="button" className={styles.action} onClick={onCopy}>
+          <DocumentDuplicateIcon className={styles.icon} />
+          <FormattedMessage id="components.common.codebox.copy" />
+        </Button>
+      </div>
+    </div>
+  );
+};
+
+export const MDXCodeBox: FC<PropsWithChildren> = ({ children: code }) => {
+  const props = isValidElement(code) ? code.props : {};
+  const { 'data-language': language } = props;
+
+  return (
+    <CodeBox language={getCodeLanguageDisplayName(language)}>{code}</CodeBox>
+  );
+};
+
+function transformCode(code: ReactNode): ReactNode {
```

Comment:
Why exactly do we need this?

Reply:
It is transforming a raw code element by breaking it down and building the content using `spans`. This way it is possible to show the line numbers.

Example:

Given the following code content:

![image](https://github.com/nodejs/nodejs.org/assets/23018799/e675afad-6041-4d3e-b170-e62c3c1880ca)

It is transformed to the following DOM structure:

![image](https://github.com/nodejs/nodejs.org/assets/23018799/ed787e6c-2990-42a7-be9a-a5d19ad0099d)

Reply:
Right that makes sense, can you make it a arrow function instead of regular function. We avoid using regular functions on TypeScript files (just for consistency)

And it'd be great if you document inline what this function does.